tutorials/examples/train_diffusion_sampler.py

```diff
@@ -0,0 +1,536 @@
+from gfn.gflownet import TBGFlowNet
+from gfn.gym.diffusion_sampling import DiffusionSampling
+from gfn.samplers import Sampler
+from gfn.states import States
+from gfn.utils.common import set_seed
+
+LOGTWOPI: float = math.log(2 * math.pi)
+
+
+##############################
+### Neural Network Modules ###
+##############################
+
+
+class PISTimeEncoding(nn.Module):
+    def __init__(self, harmonics_dim: int, t_emb_dim: int, hidden_dim: int) -> None:
+        super().__init__()
+        self.timestep_phase = nn.Parameter(torch.randn(harmonics_dim)[None])
+        self.t_model = nn.Sequential(
+            nn.Linear(2 * harmonics_dim, hidden_dim),
+            nn.GELU(),
+            nn.Linear(hidden_dim, t_emb_dim),
+        )
+        self.register_buffer(
+            "pe", torch.linspace(start=0.1, end=100, steps=harmonics_dim)[None]
+        )
+
+    def forward(self, t: torch.Tensor) -> torch.Tensor:
+        """
+        Arguments:
+            t: torch.Tensor
+        """
+        t_sin = ((t.unsqueeze(1) * self.pe) + self.timestep_phase).sin()  # type: ignore
+        t_cos = ((t.unsqueeze(1) * self.pe) + self.timestep_phase).cos()  # type: ignore
+        t_emb = torch.cat([t_sin, t_cos], dim=-1)
+        return self.t_model(t_emb)
+
+
+class PISStateEncoding(nn.Module):
+    def __init__(self, x_dim: int, s_emb_dim: int) -> None:
+        super().__init__()
+
+        self.s_model = nn.Linear(x_dim, s_emb_dim)
+
+    def forward(self, s: torch.Tensor) -> torch.Tensor:
+        return self.s_model(s)
+
+
+class PISJointPolicy(nn.Module):
+    def __init__(
+        self,
+        s_emb_dim: int,
+        hidden_dim: int,
+        out_dim: int,
+        num_layers: int,
+        zero_init: bool = False,
+    ) -> None:
+        super().__init__()
+
+        self.model = nn.Sequential(
+            nn.GELU(),
+            nn.Linear(s_emb_dim, hidden_dim),
+            nn.GELU(),
+            *[
+                nn.Sequential(nn.Linear(hidden_dim, hidden_dim), nn.GELU())
+                for _ in range(num_layers - 1)
+            ],
+            nn.Linear(hidden_dim, out_dim),
+        )
+
+        if zero_init:
+            self.model[-1].weight.data.fill_(1e-8)  # type: ignore
+            self.model[-1].bias.data.fill_(0.0)  # type: ignore
+
+    def forward(self, s_emb: torch.Tensor, t_emb: torch.Tensor) -> torch.Tensor:
+        return self.model(s_emb + t_emb)
+
+
+class PISGradNetForward(nn.Module):  # TODO: support Learnable Backward policy
+    """PISGradNet from https://arxiv.org/abs/2111.15141.
+
+    Attributes:
+    """
+
+    def __init__(
+        self,
+        s_dim: int,  # dimension of states (== target.dim)
+        harmonics_dim: int = 64,
+        t_emb_dim: int = 64,
+        s_emb_dim: int = 64,
+        hidden_dim: int = 64,
+        joint_layers: int = 2,
+        zero_init: bool = False,
+        # predict_flow: bool,  # TODO: support predict flow for db or subtb
```

Comment:
Just to bring to attention there are a bunch of TODO here
It is fine to merge them to master